File: packages/beauty_mediapipe/ios/Classes/PoseLandmarkerBridge.swift
import Foundation
import UIKit
import MediaPipeTasksVision

final class PoseLandmarkerBridge {
  private var poseLandmarker: PoseLandmarker?

  func initialize(modelPath: String) throws {
    dispose()
    guard FileManager.default.fileExists(atPath: modelPath) else {
      throw NSError(
        domain: "beauty_mediapipe",
        code: 404,
        userInfo: [NSLocalizedDescriptionKey: "Model file not found: \(modelPath)"]
      )
    }

    var options = PoseLandmarkerOptions()
    options.baseOptions.modelAssetPath = modelPath
    options.runningMode = .image
    options.numPoses = 1
    options.minPoseDetectionConfidence = 0.5
    options.minPosePresenceConfidence = 0.5
    options.minTrackingConfidence = 0.5

    poseLandmarker = try PoseLandmarker(options: options)
  }

  func detectPose(imageBytes: Data, rotation: Int) throws -> [String: Any]? {
    guard let landmarker = poseLandmarker else { return nil }
    guard let image = UIImage(data: imageBytes) else { return nil }

    let oriented = rotate(image: image, degrees: rotation)
    guard let cgImage = oriented.cgImage else { return nil }

    let mpImage = try MPImage(uiImage: UIImage(cgImage: cgImage))
    let result = try landmarker.detect(image: mpImage)

    guard let landmarks = result.landmarks.first, !landmarks.isEmpty else {
      return nil
    }

    var mappedLandmarks: [[String: Any]] = []
    var minX = 1.0, minY = 1.0, maxX = 0.0, maxY = 0.0

    for (index, landmark) in landmarks.enumerated() {
      let x = Double(landmark.x)
      let y = Double(landmark.y)
      minX = min(minX, x)
      minY = min(minY, y)
      maxX = max(maxX, x)
      maxY = max(maxY, y)
      mappedLandmarks.append([
        "index": index,
        "x": x,
        "y": y,
        "z": Double(landmark.z),
        "visibility": Double(landmark.visibility ?? 0),
      ])
    }

    return [
      "landmarks": mappedLandmarks,
      "boundingBox": [
        "left": minX,
        "top": minY,
        "right": maxX,
        "bottom": maxY,
      ],
    ]
  }

  func dispose() {
    poseLandmarker = nil
  }

  private func rotate(image: UIImage, degrees: Int) -> UIImage {
    guard degrees % 360 != 0 else { return image }
    let radians = CGFloat(degrees) * .pi / 180
    let newSize = CGRect(origin: .zero, size: image.size)
      .applying(CGAffineTransform(rotationAngle: radians))
      .integral.size

    UIGraphicsBeginImageContext(newSize)
    defer { UIGraphicsEndImageContext() }
    guard let context = UIGraphicsGetCurrentContext() else { return image }

    context.translateBy(x: newSize.width / 2, y: newSize.height / 2)
    context.rotate(by: radians)
    image.draw(
      in: CGRect(
        x: -image.size.width / 2,
        y: -image.size.height / 2,
        width: image.size.width,
        height: image.size.height
      )
    )
    return UIGraphicsGetImageFromCurrentImageContext() ?? image
  }
}

File: packages/beauty_mediapipe/android/src/main/kotlin/com/editaiapp/beauty_mediapipe/PoseLandmarkerBridge.kt
package com.editaiapp.beauty_mediapipe

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarkerResult
import java.io.File

class PoseLandmarkerBridge(private val context: Context) {
    private var poseLandmarker: PoseLandmarker? = null

    fun initialize(modelPath: String) {
        dispose()
        val modelFile = File(modelPath)
        if (!modelFile.exists()) {
            throw IllegalArgumentException("Model file not found: $modelPath")
        }

        val baseOptions = BaseOptions.builder()
            .setModelAssetPath(modelPath)
            .build()

        val options = PoseLandmarker.PoseLandmarkerOptions.builder()
            .setBaseOptions(baseOptions)
            .setRunningMode(RunningMode.IMAGE)
            .setNumPoses(1)
            .setMinPoseDetectionConfidence(0.5f)
            .setMinPosePresenceConfidence(0.5f)
            .setMinTrackingConfidence(0.5f)
            .build()

        poseLandmarker = PoseLandmarker.createFromOptions(context, options)
    }

    fun detectPose(imageBytes: ByteArray, rotation: Int): Map<String, Any?>? {
        val landmarker = poseLandmarker ?: return null

        val bitmap = decodeBitmap(imageBytes, rotation) ?: return null
        val mpImage = BitmapImageBuilder(bitmap).build()

        return try {
            val result: PoseLandmarkerResult = landmarker.detect(mpImage)
            mapResult(result)
        } catch (e: Exception) {
            Log.e(TAG, "detectPose failed", e)
            null
        } finally {
            if (!bitmap.isRecycled) {
                bitmap.recycle()
            }
        }
    }

    fun dispose() {
        poseLandmarker?.close()
        poseLandmarker = null
    }

    private fun decodeBitmap(imageBytes: ByteArray, rotation: Int): Bitmap? {
        val options = BitmapFactory.Options().apply {
            inPreferredConfig = Bitmap.Config.ARGB_8888
        }
        var bitmap = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size, options)
            ?: return null

        if (rotation != 0) {
            val matrix = android.graphics.Matrix().apply {
                postRotate(rotation.toFloat())
            }
            val rotated = Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
            if (rotated != bitmap) {
                bitmap.recycle()
            }
            bitmap = rotated
        }
        return bitmap
    }

    private fun mapResult(result: PoseLandmarkerResult): Map<String, Any?>? {
        if (result.landmarks().isEmpty()) {
            return null
        }

        val landmarks = result.landmarks()[0]
        if (landmarks.isEmpty()) {
            return null
        }

        val mappedLandmarks = ArrayList<Map<String, Any>>(landmarks.size)
        var minX = 1.0
        var minY = 1.0
        var maxX = 0.0
        var maxY = 0.0

        landmarks.forEachIndexed { index, landmark ->
            val x = landmark.x().toDouble()
            val y = landmark.y().toDouble()
            minX = minOf(minX, x)
            minY = minOf(minY, y)
            maxX = maxOf(maxX, x)
            maxY = maxOf(maxY, y)
            mappedLandmarks.add(
                mapOf(
                    "index" to index,
                    "x" to x,
                    "y" to y,
                    "z" to landmark.z().toDouble(),
                    "visibility" to landmark.visibility().orElse(0.0f).toDouble(),
                ),
            )
        }

        return mapOf(
            "landmarks" to mappedLandmarks,
            "boundingBox" to mapOf(
                "left" to minX,
                "top" to minY,
                "right" to maxX,
                "bottom" to maxY,
            ),
        )
    }

    companion object {
        private const val TAG = "PoseLandmarkerBridge"
    }
}
